### PR TITLE
Patch binary code with NOP or NULL bytes.

### PR DESCRIPTION
**Pull request by quangnh89**
This feature is very similar to  OllyDbg.

example

### CODE REVIEW

**Comment by redragonvn** [comment]
Your "Fill with NOPs" is x86-specific code. Please make sure that "Fill with NOPs" context menu is only registered on supported arch (x86).

**Comment by quangnh89** [comment]
Thanks for review my work. I have fixed this issue and committed code.

**Comment by aquynh** [comment]
this is nice, but to be consistent with existing features, there are still some works to do:

- can you add **IDA comment** to where you patch, in the form "Keypatch modified this ...".?
- the current "Undo" function should be able to revert this patch

see related screenshots here https://twitter.com/keystone_engine/status/764847048344477696

File: keypatch.py [hunk 1]
```diff
@@ -1109,6 +1109,78 @@ def OnFormChange(self, fid):
         return 1
 
 
+class Binary_Fill_NOPs(idaapi.action_handler_t):
```
Comment (aquynh) [inline keypatch.py]: the way you reuse this `Binary_Fill_NOPs` class for `Binary_Fill_NULLs` is a bit weird. i would do that differently: rename `Binary_Fill_NOPs` to `Binary_Fill`, and pass `fill_value` (either `0x90` or `0x00`) & label (either "Fill with 00's" or "Fill with NOPs") to the class constructor.

Comment (quangnh89) [inline keypatch.py]: Thank you. I will take your suggestion.

**Comment by balika011** [comment]
This looks usefull. What about make it support any arch? (Like ARM.)

**Comment by quangnh89** [comment]
@balika011 Thank for your advice. I will make it support any arch.
I think the feature which re-analyzes automatically after patching makes me slow-down. "reanalyse" option is needed.

reanalyze

``` python
def patch_code(self, address, assembly, syntax, padding, save_origcode, reanalyze = True, orig_asm = None, patch_data = None):
        global patch_info
# ........
 if reanalyze:
    # ask IDA to re-analyze the patched area
    idaapi.analyze_area(address, orig_func_end)

    # try to fix IDA function re-analyze issue after patching
    idaapi.func_setend(address, orig_func_end)
```

**Comment by aquynh** [comment]
Reanalyzing should be a must, but not a choice. This is what we are already
doing with the current patcher.

**Comment by quangnh89** [comment]
I'm analyzing a malware obfuscated by some tools and analysis of IDA always fails to detect address of the end of function. I can send you this sample for further information.

**Comment by aquynh** [comment]
In that case, we can still go ahead patch ing, but do not reanalyze code.
But we should log this issue, and warn user on what happened (using
Warning() popup message, for example).

If you have a better idea, lets discuss here.

**Comment by aquynh** [comment]
merged into the "test" branch at https://github.com/keystone-engine/keypatch/tree/test
will merge into "master" when it is ready, thanks.

**Comment by aquynh** [comment]
please send future pull requests for the "test" branch.

**Pull request closed without merging**